### ISSUE TITLE
MultiWorkerMirroredStrategy reinitializes reduce operations multiple times at start and also from time to time.

### ISSUE THREAD

**Issue by andre-bauer** [tensorflow/tensorflow#55147]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: -
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.7.0-rc1-69-gc256c071bb2 2.7.0
- Python version: 3.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 11.4 
- GPU model and memory: V100 16GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Fyi: I use a custom training loop. 
On a multi worker setup with `MultiWorkerMirroredStrategy` I get this output for the first 3 train steps and they are also super slow:
```
[08/03/22 16:50:11] [tensorflow] INFO : Collective all_reduce tensors: 236 all_reduces, num_devices = 8, group_size = 48, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 8, group_size = 48, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
[08/03/22 16:50:20] [tensorflow] INFO : Collective all_reduce tensors: 1 all_reduces, num_devices = 1, group_size = 6, implementation = NCCL, num_packs = 1
``` 

Then the training runs as expected but it seems that the above init/retracing can happen randomly in the middle of the training again with the same logs. So an intermediate step after 1000 iterations can again take minutes (the more workers/devices the longer)
The step at which this happens is not the same for multiple runs, that being the reason why I say randomly.

Could this be related to non-constant input sizes (number of points in a pointcloud? They are of course padded within a batch) Do I need to globally pad here?

**Describe the expected behavior**
Only initialize the Collective all_reduce tensors once and especially not in the middle of training.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by chunduriv** [comment]
@andre-bauer,

In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by andre-bauer** [comment]
@chunduriv Thx for the reply I actually figured it out myself and it was my fault. Still documenting the problem I created here:

This out above happens either on model build or on function tracing. Apparently I was passing some python object into the tf.function that could change from time to time cause function retracing. With one worker there is no output (default logging verbosity, relax_shapes=True) and it also does not take a crazy amount of time. That is the reason why I only noticed it when using ``MultiWorkerMirroredStrategy``. So to everyone having something similar, check if your tf.function is retraced for some reason.
